File: big-o-notation/exercises/is_palindrome/is_palindrome.py
def is_palindrome(s: str) -> bool:
    """
    Checks if a string is a palindrome using Python's slicing.
    Time Complexity: O(n)
    Space Complexity: O(n) (due to string creation during slicing)
    """
    # Normalize
    s = s.lower().replace(" ", "")
    return s == s[::-1]

def is_palindrome_optimized(s: str) -> bool:
    """
    Checks if a string is a palindrome using two pointers.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    s = s.lower().replace(" ", "")
    left, right = 0, len(s) - 1
    while left < right:
        if s[left] != s[right]:
            return false
        left += 1
        right -= 1
    return True

if __name__ == "__main__":
    test_cases = ["Racecar", "hello", "A man a plan a canal Panama"]
    for tc in test_cases:
        print(f"Is '{tc}' a palindrome? {is_palindrome(tc)}")

File: big-o-notation/exercises/is_palindrome/is_palindrome.ts
// is_palindrome.ts

/**
 * Checks if a string is a palindrome.
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
export function isPalindrome(s: string): boolean {
    const normalized = s.toLowerCase().replace(/[^a-z0-9]/g, "");
    const reversed = normalized.split("").reverse().join("");
    return normalized === reversed;
}

/**
 * Optimized Two-Pointer approach.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
export function isPalindromeOptimized(s: string): boolean {
    const normalized = s.toLowerCase().replace(/[^a-z0-9]/g, "");
    let left = 0;
    let right = normalized.length - 1;

    while (left < right) {
        if (normalized[left] !== normalized[right]) {
            return false;
        }
        left++;
        right--;
    }
    return true;
}

// Testing
const testCases = ["Racecar", "hello", "A man, a plan, a canal: Panama"];
testCases.forEach((tc) => {
    console.log(`Is "${tc}" a palindrome? ${isPalindromeOptimized(tc)}`);
});
